src/System.Text.Primitives/System/Text/Encoders/Utf8Util.Validation.cs

```diff
@@ -0,0 +1,728 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+
+//
+// This file contains workhorse methods for performing validation of UTF-8 byte sequences.
+//
+
+namespace System.Buffers.Text
+{
+    internal static partial class Utf8Util
+    {
+        // This method will consume as many ASCII bytes as it can using fast vectorized processing, returning the number of
+        // consumed (ASCII) bytes. It's possible that the method exits early, perhaps because there is some non-ASCII byte
+        // later in the sequence or because we're running out of input to search. The intent is that the caller *skips over*
+        // the number of bytes returned by this method, then it continues data processing from the next byte.
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static unsafe int ConsumeAsciiBytesVectorized(ref byte buffer, int length)
+        {
+            // Only allow vectorization if vector operations are hardware-accelerated
+            // and vectors are at least 128 bits in length.
+            if (Vector.IsHardwareAccelerated && (Vector<byte>.Count >= 2 * sizeof(ulong)))
```

Comment:
Should you skip vectorized implementation if the length is small?

Reply:
Yes, this is checked both by the caller (to avoid the non-inlineable call) and again by this method on line 37.